docs: integrating Formik into Carbon FE-2283 (#2481)
Sage/carbon: rfcs/text/formik.md

itself has 8 subdependencies.  Carbon is already using 4 of those subdependencies, so 4 additional subdependencies would be installed
([`deepmerge`](https://www.npmjs.com/package/deepmerge),
[`hoist-non-react-statics`](https://www.npmjs.com/package/hoist-non-react-statics),
[`scheduler`](https://www.npmjs.com/package/scheduler) and
[`tiny-warning`](https://www.npmjs.com/package/tiny-warning)).  If we wish to also install
[`yup`](https://www.npmjs.com/package/yup) for schema-based validation, this would add another 4 subdependencies
([`fn-name`](https://www.npmjs.com/package/fn-name),
[`property-expr`](https://www.npmjs.com/package/property-expr),
[`synchronous-promise`](https://www.npmjs.com/package/synchronous-promise) and
[`toposort`](https://www.npmjs.com/package/toposort)).

Some of the key disadvantages/risks of using a third-party library are:

* **Risk of abandonment**  –  It's very common for open-source projects to eventually be abandoned by their
  maintainers.  When this happens, any current or future bugs will never be fixed, and community-provided support (on

>   GitHub, Stack Overflow, etc) will gradually decline.

Reviewer: It may be useful to ask a question along the lines of 

> would we be comfortable maintaining the library ourselves if the current maintainer stopped doing so and community support failed to fill the void?

Reviewer: This of course is no worse than our current situation of maintaining our own attempt within Carbon.

Reviewer 2: Marking as resolved. I don't want the RFC to introduce more questions, I'm happy with the statement as it stands.

* **Familiarity**  –  Users are potentially already familiar with popular third-party libraries.

Considering the reasons why "reinventing the wheel" (ie: building a custom solution) can sometimes be beneficial in
certain situations:

* It can be a very educational experience for developers, but that isn't a relevant factor for this project.
* A third-party library may be an unnecessarily heavy solution, if only a small percentage of its functionality will be
  used by the application, and a custom-built solution may be much more lightweight.  But in this case, Formik is
  [a relatively small library](https://jaredpalmer.com/formik/docs/overview)
  ([14.7kB when minified and gzipped](https://bundlephobia.com/result?p=formik@2.0.4)), and we would be using a
  significant percentage of its functionality.

# Detailed design

## Overview of Carbon's form and validation functionality

Carbon provides the following components:

* [**`<Form>`**](src/__experimental__/components/form/form.component.js)  –  This augments an HTML
  [`<form>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form) element with: a footer (containing "Save"

>   and "Cancel" buttons), form submission functionality, [CRSF](https://en.wikipedia.org/wiki/CRSF) functionality, a

Reviewer 3: Typo CRSF -> CSRF

Author: Wow, well spotted, thanks! :clap: Fixed in db9e307

  [`beforeunload`](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event) "unsaved" warning, etc.
* [**`<FormField>`**](src/__experimental__/components/form-field/form-field.component.js)  –  This enables a
  [`<Label>`](src/__experimental__/components/label/label.component.js) and
  [`<FieldHelp>`](src/__experimental__/components/field-help/field-help.component.js) to be added to an individual
  form field.
* [**`withValidation`**](src/components/validations/with-validation.hoc.js)  –  This HOC wraps an individual form field,
  enabling validation functions (for errors, warnings and info) to be executed when the component blurs or receives
  updated props.

## Overview of Formik

> Formik keeps track of your form's state, and then exposes it (plus a few reusable methods and event handlers) to your
> form via props. [🔗](https://jaredpalmer.com/formik/docs/overview#the-gist)

Formik provides a small set of React components and hooks, of which the key ones are:

* [**`<Formik>`**](https://jaredpalmer.com/formik/docs/api/formik)  –  The central component of Formik, which wraps an
  HTML form, and accepts props such as

>   [`initialValues`](https://jaredpalmer.com/formik/docs/api/formik#initialvalues-values),

Reviewer 3: We will probably want to document somewhere that Formik will only work for controlled inputs, highly unlikely that someone would try to use it with uncontrolled but better if we make it explicit IMO

Author: Good point :thumbsup: I suppose technically it's possible to use Formik with an uncontrolled input, by wiring up [`handleChange`](https://jaredpalmer.com/formik/docs/api/formik#handlechange-e-reactchangeevent-any-void) (for Formik's validation functionality) but not wiring up `value` to Formik. But that would cause complications, and it's not really how Formik is intended to be used.

Reviewer 3: Yeah good point, think we are best not encouraging it's use uncontrolled but we can't really force the hand of developers using Carbon

Author: Added documentation about uncontrolled components in 3516b7f :thumbsup:

- Start Date: 2019-11-15

# Table of contents

- [Table of contents](#table-of-contents)
- [Summary](#summary)
- [Motivation](#motivation)
- [Detailed design](#detailed-design)
  - [Overview of Carbon's form and validation functionality](#overview-of-carbons-form-and-validation-functionality)
  - [Overview of Formik](#overview-of-formik)
    - [Form validation using Formik](#form-validation-using-formik)
  - [Code examples](#code-examples)
    - [Checkbox and CheckboxGroup](#checkbox-and-checkboxgroup)
      - [HTML and React](#html-and-react)
      - [Checkbox usage with Formik](#checkbox-usage-with-formik)
      - [Form-level validation examples](#form-level-validation-examples)
      - [Simplification using `<Field>`](#simplification-using-field)
      - [Field-level validation example](#field-level-validation-example)
      - [CheckboxGroup example: uncontrolled](#checkboxgroup-example-uncontrolled)
      - [CheckboxGroup example: controlled](#checkboxgroup-example-controlled)
    - [Blocking navigation on unsaved changes](#blocking-navigation-on-unsaved-changes)
- [Drawbacks](#drawbacks)

>   - [No support for non-blocking ("warning" or "info") validations](#no-support-for-non-blocking-%22warning%22-or-%22info%22-validations)

Author: @edleeks87: This change you made in 6f93191 seems to have broken the link:

```diff
- [No support for non-blocking ("warning" or "info") validations](#no-support-for-non-blocking-warning-or-info-validations)
+ [No support for non-blocking ("warning" or "info") validations](#no-support-for-non-blocking-%22warning%22-or-%22info%22-validations)
```

This link doesn't work for me:

* https://github.com/Sage/carbon/blob/FE-2283-formik-rfc/rfcs/text/formik.md#no-support-for-non-blocking-%22warning%22-or-%22info%22-validations

But the original link works for me:

* https://github.com/Sage/carbon/blob/FE-2283-formik-rfc/rfcs/text/formik.md#no-support-for-non-blocking-warning-or-info-validations

Please can we revert the change? :point_down:

```suggestion
  - [No support for non-blocking ("warning" or "info") validations](#no-support-for-non-blocking-warning-or-info-validations)
```

Reviewer 2: docs: revert link format

- Start Date: 2019-11-15

# Table of contents

- [Table of contents](#table-of-contents)
- [Summary](#summary)
- [Motivation](#motivation)
- [Detailed design](#detailed-design)
  - [Overview of Carbon's form and validation functionality](#overview-of-carbons-form-and-validation-functionality)
  - [Overview of Formik](#overview-of-formik)
    - [Form validation using Formik](#form-validation-using-formik)
  - [Code examples](#code-examples)
    - [Checkbox and CheckboxGroup](#checkbox-and-checkboxgroup)
      - [HTML and React](#html-and-react)
      - [Checkbox usage with Formik](#checkbox-usage-with-formik)
      - [Form-level validation examples](#form-level-validation-examples)
      - [Simplification using `<Field>`](#simplification-using-field)
      - [Field-level validation example](#field-level-validation-example)
      - [CheckboxGroup example: uncontrolled](#checkboxgroup-example-uncontrolled)
      - [CheckboxGroup example: controlled](#checkboxgroup-example-controlled)
    - [Blocking navigation on unsaved changes](#blocking-navigation-on-unsaved-changes)
- [Drawbacks](#drawbacks)

>   - [No support for non-blocking ("warning" or "info") validations](#no-support-for-non-blocking-%22warning%22-or-%22info%22-validations)

Reviewer: I think the entity characters are screwing with the skip-links here

Reviewer 3: @ianforster-sage  not sure what you mean, the links seem to navigate to the correct sections and the text seems fine, could you clarify for me as I may be missing something?

          </Field>
          <div style={{ color: 'red'    }}>{errors.shape}</div>
          <div style={{ color: 'orange' }}>{shapeWarning}</div>
          <div style={{ color: 'blue'   }}>{shapeInfo}</div>
          <input type="submit" />
        </Form>
      )}
    </Formik>
  );

};

render(<App />, document.getElementById('root'));
```

# Alternatives

There are a number of alternative packages available to developers looking for a solution to web forms and validation. It is our intention to provide an API that affords developers the flexibility to choose whatever solution they want to and integrate their choice with Carbon's Form and input components. 

The alternative to integrating Formik is to continue to maintain the current implementation of Form and Validations in Carbon. We would then continue to add additional functionality to meet any new requirements that may arise. This currently already adds a significant overhead to the team any will only grow as new functionality is implemented. A further issue is that it will continue to cause vendor lock-in, forcing consumers to use what is provided by Carbon out-of-the-box, and may hurt its adoption as a result.

> # Adoption strategy

Reviewer 4: We will need to think about what this looks like to migrate from how forms are built today - to how forms are built tomorrow using Formik. I.e. if i'm a developer with a project using v12.0.0 of Carbon, what do i need to do to upgrade to the version with Formik support.

> As discussed, we propose to remove the majority of the code from the `Carbon` `Form` and have it serve predominantly as a UI wrapper that renders either a `Formik` `Form` or an alterantive package chosen by the developers working on the consuming projects. Carbon's From will no longer be responsible for implementing the valididation interface that it currently does: the Form component will no longer be exported with the functionality added by the `withValidationHOC`.

Reviewer 4: ```suggestion
As discussed, we propose to remove the majority of the code from the `Carbon` `Form` and have it serve predominantly as a UI wrapper that renders either a `Formik` `Form` or an alternative package chosen by the developers working on the consuming projects. Carbon's Form will no longer be responsible for implementing the validation interface that it currently does: the Form component will no longer be exported with the functionality added by the `withValidationHOC`.
```

> As such, thevalidation interface will need to be simplified and updated as the use of the `withValidationHOC` is currently required to update the `FormSummary` sub-component and control the acceptance criteria for a successful form submission via a successful run of the `validate` function. The removal of a key aspect of the current validation API means carrying out the changes proposed in this RFC would be a breaking change.

Reviewer 4: ```suggestion
As such, the validation interface will need to be simplified and updated as the use of the `withValidationHOC` is currently required to update the `FormSummary` sub-component and control the acceptance criteria for a successful form submission via a successful run of the `validate` function. The removal of a key aspect of the current validation API means carrying out the changes proposed in this RFC would be a breaking change.
```
